test/extended/util/client.go

```diff
@@ -770,52 +812,60 @@ func (c *CLI) CreateUser(prefix string) *userv1.User {
-			ObjectMeta: metav1.ObjectMeta{Name: username},
-		}, metav1.CreateOptions{})
-		if err != nil {
+	if userAPIExists {
+		userClientConfig = rest.AnonymousClientConfig(turnOffRateLimiting(rest.CopyConfig(c.AdminConfig())))
+		ctx := context.Background()
+		userClient := c.AdminUserClient()
+		user, err := userClient.UserV1().Users().Get(ctx, username, metav1.GetOptions{})
+		if err != nil && !apierrors.IsNotFound(err) {
 			FatalErr(err)
 		}
-		c.AddResourceToDelete(userv1.GroupVersion.WithResource("users"), user)
-	}
+		if err != nil {
+			user, err = userClient.UserV1().Users().Create(ctx, &userv1.User{
+				ObjectMeta: metav1.ObjectMeta{Name: username},
+			}, metav1.CreateOptions{})
+			if err != nil {
+				FatalErr(err)
+			}
+			c.AddResourceToDelete(userv1.GroupVersion.WithResource("users"), user)
+		}
 
-	oauthClient := c.AdminOauthClient()
-	oauthClientName := "e2e-client-" + c.Namespace()
-	oauthClientObj, err := oauthClient.OauthV1().OAuthClients().Create(ctx, &oauthv1.OAuthClient{
-		ObjectMeta:  metav1.ObjectMeta{Name: oauthClientName},
-		GrantMethod: oauthv1.GrantHandlerAuto,
-	}, metav1.CreateOptions{})
-	if err != nil && !apierrors.IsAlreadyExists(err) {
-		FatalErr(err)
-	}
-	if oauthClientObj != nil {
-		c.AddExplicitResourceToDelete(oauthv1.GroupVersion.WithResource("oauthclients"), "", oauthClientName)
-	}
+		oauthClient := c.AdminOauthClient()
+		oauthClientName := "e2e-client-" + c.Namespace()
+		oauthClientObj, err := oauthClient.OauthV1().OAuthClients().Create(ctx, &oauthv1.OAuthClient{
+			ObjectMeta:  metav1.ObjectMeta{Name: oauthClientName},
+			GrantMethod: oauthv1.GrantHandlerAuto,
+		}, metav1.CreateOptions{})
+		if err != nil && !apierrors.IsAlreadyExists(err) {
+			FatalErr(err)
+		}
+		if oauthClientObj != nil {
+			c.AddExplicitResourceToDelete(oauthv1.GroupVersion.WithResource("oauthclients"), "", oauthClientName)
+		}
 
-	privToken, pubToken := GenerateOAuthTokenPair()
-	token, err := oauthClient.OauthV1().OAuthAccessTokens().Create(ctx, &oauthv1.OAuthAccessToken{
-		ObjectMeta:  metav1.ObjectMeta{Name: pubToken},
-		ClientName:  oauthClientName,
-		UserName:    username,
-		UserUID:     string(user.UID),
-		Scopes:      []string{"user:full"},
-		RedirectURI: "https://localhost:8443/oauth/token/implicit",
-	}, metav1.CreateOptions{})
-	if err != nil {
-		FatalErr(err)
+		privToken, pubToken := GenerateOAuthTokenPair()
+		token, err := oauthClient.OauthV1().OAuthAccessTokens().Create(ctx, &oauthv1.OAuthAccessToken{
+			ObjectMeta:  metav1.ObjectMeta{Name: pubToken},
+			ClientName:  oauthClientName,
+			UserName:    username,
+			UserUID:     string(user.UID),
+			Scopes:      []string{"user:full"},
+			RedirectURI: "https://localhost:8443/oauth/token/implicit",
+		}, metav1.CreateOptions{})
+		if err != nil {
+			FatalErr(err)
+		}
+		c.AddResourceToDelete(oauthv1.GroupVersion.WithResource("oauthaccesstokens"), token)
+		userClientConfig.BearerToken = privToken
+	} else {
```

Comment:
The change in this function would be smaller if the `if` test on line 821 was reversed and then the userClientConfig being created here was returned early. That would also make the function easier to follow, because it would be clear that most of the steps apply when the users API is present.

Reply:
You are right, looks better now. Thanks!